File: Medium/Stickler Thief/stickler-thief.py
#User function Template for python3

class Solution:  
    
    #Function to find the maximum money the thief can get.
    def FindMaxSum(self,a, n):
        
        # code here
        if n==0 :
            return 0
        if n==1 :
            return a[0]
        dp=[0]*(n)
        dp[0]=a[0]
        dp[1]=max(a[0],a[1])
        for i in range(2,n) :
            dp[i] = max(dp[i-1], dp[i-2]+a[i])
        return dp[n-1]
        
        
        
#class Solution:  
    
    #Function to find the maximum money the thief can get.
#    def FindMaxSum(self,a, n):
        
        # code here
#        def rec(i,ans,dp):
            
#            if i >= n:
#                return ans
#            if (i,ans) in dp :
#                return dp[(i,ans)]
                
#            take = rec(i + 2, ans + a[i],dp)
#            nottake = rec(i + 1, ans,dp)
#            dp[(i,ans)]=max(take, nottake)
            
#            return dp[(i,ans)]
            
#        dp=dict()
#        return rec(0,0,dp)





#{ 
 # Driver Code Starts
#Initial Template for Python 3
import atexit
import io
import sys
sys.setrecursionlimit(10**6)
# Contributed by : Nagendra Jha

if __name__ == '__main__':
    test_cases = int(input())
    for cases in range(test_cases):
        n = int(input())
        a = list(map(int,input().strip().split()))
        ob=Solution()
        print(ob.FindMaxSum(a,n))
# } Driver Code Ends

File: Medium/Stickler Thief/stickler-thief.java
//{ Driver Code Starts
import java.util.*;
import java.io.*;

class GFG
 {
	public static void main (String[] args)
	 {
	  
	  //taking input using Scanner class
	  Scanner sc = new Scanner(System.in);
	  
	  //taking count of testcases
	  int t = sc.nextInt();
	  while(t-- > 0){
	      
	      //taking count of houses
	      int n = sc.nextInt();
	      int arr[] = new int[n];
	      
	      //inserting money present in 
	      //each house in the array
	      for(int i = 0; i<n; ++i)
	           arr[i] = sc.nextInt();
  	      
  	      //calling method FindMaxSum() of class solve
  	      System.out.println(new Solution().FindMaxSum(arr, n));
	 }
   }
}
// } Driver Code Ends


class Solution
{
    //Function to find the maximum money the thief can get.
    public int FindMaxSum(int[] a, int n) {
        // Create an array for memoization
        int[] dp = new int[n];
        Arrays.fill(dp, -1); // Initialize the array with -1 to indicate uninitialized values
        return rec(0, a, n, dp);
    }
    
    // Recursive function with memoization
    private int rec(int i, int[] a, int n, int[] dp) {
        if (i >= n) {
            return 0;
        }
        
        // Check if the result is already memoized
        if (dp[i] != -1) {
            return dp[i];
        }
                
        int take = a[i] + rec(i + 2, a, n, dp);
        int notTake = rec(i + 1, a, n, dp);
        
        // Store the result in the memoization array
        dp[i] = Math.max(take, notTake);
        
        return dp[i];
    }
}